File: LLCycle.java
import structs.ListNode;

public class LLCycle {
  public static void main(String[] args){
    LLCycle s = new LLCycle();
    ListNode node2 = new ListNode(2);
    ListNode head = new ListNode(3, node2);
    node2.next = new ListNode(0, new ListNode(4, node2));
    System.out.println(s.hasCycle(head));
  }

  public boolean hasCycle(ListNode head) {
    ListNode slow = head, fast = head;
    while(fast != null && fast.next != null){
      slow = slow.next;
      fast = fast.next.next;
      if(slow == fast){
        return true;
      }
    }
    return false;
  }
}

File: LinkedList/ll_cycle.go
package main

import "fmt"

type ListNode struct {
    Val int
    Next *ListNode
}

func hasCycle(head *ListNode) bool {
    slow, fast := head, head
	for fast != nil && fast.Next != nil{
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast{
			return true
		}
	}
	return false
}

func main(){
	node3 := &ListNode{4, nil}
	node2 := &ListNode{2, &ListNode{0, node3}}
	node1 := &ListNode{3, node2}
	node3.Next = node2
	fmt.Println(hasCycle(node1))
}